File: packages/stores/src/dynamodb.ts
import type {
  BeginOptions,
  BeginOutcome,
  CompleteStatus,
  IdempotencyRecord,
  OmittedResult,
  Operation,
  Store,
  StoredResult,
} from '@anyonce/core';
import { isOmitted } from '@anyonce/core';
import {
  type AttributeValue,
  ConditionalCheckFailedException,
  CreateTableCommand,
  DeleteItemCommand,
  DescribeTableCommand,
  type DynamoDBClient,
  GetItemCommand,
  ResourceInUseException,
  UpdateItemCommand,
  UpdateTimeToLiveCommand,
} from '@aws-sdk/client-dynamodb';
import { encodeResultMeta, type RecordRow, rowToRecord } from './codec';

/** Q20: a DynamoDB item is capped at 400 KB, so bodies above this are stored in the omitted form. */
export const DYNAMODB_MAX_RESULT_BYTES = 307_200;

/** Unit separator: the scope and key are joined with it so one composed partition key stays unambiguous. */
const KEY_SEPARATOR = String.fromCharCode(31);

/** Q22: the item key is the single partition key pk, the scope and the key joined by the unit separator. */
export function itemKey(op: Pick<Operation, 'scope' | 'key'>): string {
  return `${op.scope}${KEY_SEPARATOR}${op.key}`;
}

export interface DynamoDbStoreOptions {
  client: DynamoDBClient;
  /** Default anyonce_records. One partition key, pk, holding the scope and the key (Q22). No sort key. */
  tableName?: string;
  /** Added to the native ttl attribute so a late complete from the previous fence holder still finds its row. Default 60000. */
  nativeTtlGraceMs?: number;
}

type Item = Record<string, AttributeValue>;

function n(value: number): AttributeValue {
  return { N: String(value) };
}

function itemToRow(item: Item): RecordRow {
  const pk = item.pk?.S ?? '';
  // Split on the first separator: a scope never contains one, a key may.
  const at = pk.indexOf(KEY_SEPARATOR);
  return {
    scope: at === -1 ? pk : pk.slice(0, at),
    key: at === -1 ? '' : pk.slice(at + 1),
    fingerprint: item.fingerprint?.S ?? '',
    state: (item.state?.S as RecordRow['state']) ?? 'in_flight',
    fence: Number(item.fence?.N ?? 0),
    lease_until: Number(item.lease_until?.N ?? 0),
    created_at: Number(item.created_at?.N ?? 0),
    expires_at: Number(item.expires_at?.N ?? 0),
    result_meta: item.result_meta?.S ?? null,
    result_body: item.result_body?.B ?? null,
    result_omitted: Number(item.result_omitted?.N ?? 0),
  };
}

/**
 * REQ-ST-DDB-1. begin, complete and abandon are each one conditional write; a refused write carries the old item
 * back through ReturnValuesOnConditionCheckFailure, so no second read is needed to classify it.
 */
export class DynamoDbStore implements Store {
  private readonly client: DynamoDBClient;
  private readonly table: string;
  private readonly grace: number;

  constructor(options: DynamoDbStoreOptions) {
    this.client = options.client;
    this.table = options.tableName ?? 'anyonce_records';
    this.grace = options.nativeTtlGraceMs ?? 60_000;
  }

  private key(op: Pick<Operation, 'scope' | 'key'>): Item {
    return { pk: { S: itemKey(op) } };
  }

  async begin(op: Operation, opts: BeginOptions): Promise<BeginOutcome> {
    for (let attempt = 0; attempt < 3; attempt++) {
      const ttlSeconds = Math.floor((Date.now() + opts.ttlMs + this.grace) / 1000);
      try {
        const out = await this.client.send(
          new UpdateItemCommand({
            TableName: this.table,
            Key: this.key(op),
            ConditionExpression:
              'attribute_not_exists(pk) OR expires_at <= :now OR (fingerprint = :fp AND #state = :in_flight AND lease_until <= :now)',
            UpdateExpression:
              'SET fingerprint = :fp, #state = :in_flight, fence = if_not_exists(fence, :zero) + :one, lease_until = :lease, created_at = :now, expires_at = :exp, #ttl = :ttl, result_omitted = :zero REMOVE result_meta, result_body',
            ExpressionAttributeNames: { '#state': 'state', '#ttl': 'ttl' },
            ExpressionAttributeValues: {
              ':fp': { S: op.fingerprint },
              ':in_flight': { S: 'in_flight' },
              ':now': n(opts.now),
              ':zero': n(0),
              ':one': n(1),
              ':lease': n(opts.now + opts.leaseMs),
              ':exp': n(opts.now + opts.ttlMs),
              ':ttl': n(ttlSeconds),
            },
            ReturnValues: 'ALL_NEW',
            ReturnValuesOnConditionCheckFailure: 'ALL_OLD',
          }),
        );
        return { outcome: 'acquired', fence: Number(out.Attributes?.fence?.N) };
      } catch (error) {
        if (!(error instanceof ConditionalCheckFailedException)) throw error;
        if (error.Item === undefined)
          throw new Error(
            'anyonce: dynamodb refused the begin write without returning the old item, so the refusal cannot be classified; the endpoint must honour ReturnValuesOnConditionCheckFailure (DynamoDB Local 2.x or later, or a live table)',
            { cause: error },
          );
        const row = itemToRow(error.Item as Item);
        if (row.expires_at <= opts.now) continue;
        const record = rowToRecord(row);
        if (row.fingerprint !== op.fingerprint) return { outcome: 'mismatch', record };
        if (row.state === 'completed') return { outcome: 'completed', record };
        if (row.lease_until > opts.now)
          return { outcome: 'in_flight', leaseUntil: row.lease_until };
      }
    }
    throw new Error(
      'anyonce: dynamodb begin could not settle after three attempts; the row for this scope and key expired or was rewritten between every conditional write and its classification, or the stored row is malformed (expires_at not a number)',
    );
  }

  async complete(
    op: Operation,
    fence: number,
    result: StoredResult | OmittedResult,
    now: number,
  ): Promise<CompleteStatus> {
    const omitted = isOmitted(result);
    const body = omitted ? undefined : result.body;
    if (body !== undefined && body.byteLength > DYNAMODB_MAX_RESULT_BYTES)
      throw new Error(
        `anyonce: dynamodb cannot store a ${body.byteLength} byte body because an item is capped at 400 KB (Q20); set maxResultBytes to at most ${DYNAMODB_MAX_RESULT_BYTES} so a larger result is stored in the omitted form instead`,
      );
    const values: Item = {
      ':fence': n(fence),
      ':now': n(now),
      ':in_flight': { S: 'in_flight' },
      ':completed': { S: 'completed' },
      ':meta': { S: encodeResultMeta(result) },
      ':om': n(omitted ? 1 : 0),
    };
    let update = 'SET #state = :completed, result_meta = :meta, result_omitted = :om';
    if (body !== undefined) {
      values[':body'] = { B: body };
      update += ', result_body = :body';
    } else {
      update += ' REMOVE result_body';
    }
    try {
      await this.client.send(
        new UpdateItemCommand({
          TableName: this.table,
          Key: this.key(op),
          ConditionExpression: 'fence = :fence AND expires_at > :now AND #state = :in_flight',
          UpdateExpression: update,
          ExpressionAttributeNames: { '#state': 'state' },
          ExpressionAttributeValues: values,
          ReturnValuesOnConditionCheckFailure: 'ALL_OLD',
        }),
      );
      return 'ok';
    } catch (error) {
      if (!(error instanceof ConditionalCheckFailedException)) throw error;
      if (error.Item === undefined) return 'not_found';
      const row = itemToRow(error.Item as Item);
      if (row.expires_at <= now) return 'not_found';
      if (row.fence !== fence) return 'stale_fence';
      return 'ok';
    }
  }

  async abandon(op: Operation, fence: number): Promise<CompleteStatus> {
    try {
      await this.client.send(
        new DeleteItemCommand({
          TableName: this.table,
          Key: this.key(op),
          ConditionExpression: 'fence = :fence AND #state = :in_flight',
          ExpressionAttributeNames: { '#state': 'state' },
          ExpressionAttributeValues: { ':fence': n(fence), ':in_flight': { S: 'in_flight' } },
          ReturnValuesOnConditionCheckFailure: 'ALL_OLD',
        }),
      );
      return 'ok';
    } catch (error) {
      if (!(error instanceof ConditionalCheckFailedException)) throw error;
      if (error.Item === undefined) return 'not_found';
      const row = itemToRow(error.Item as Item);
      if (row.state !== 'in_flight') return 'not_found';
      return row.fence === fence ? 'not_found' : 'stale_fence';
    }
  }

  async get(op: Pick<Operation, 'scope' | 'key'>, now: number): Promise<IdempotencyRecord | null> {
    const out = await this.client.send(
      new GetItemCommand({ TableName: this.table, Key: this.key(op), ConsistentRead: true }),
    );
    if (out.Item === undefined) return null;
    const row = itemToRow(out.Item as Item);
    return row.expires_at > now ? rowToRecord(row) : null;
  }

  /** Native TTL sweeps expired items; nothing to do here (REQ-ST-DDB-1). */
  async purge(_now: number): Promise<number> {
    return 0;
  }

  /** Test-only: what a TTL sweep would do. */
  async physicallyRemove(op: Pick<Operation, 'scope' | 'key'>): Promise<void> {
    await this.client.send(new DeleteItemCommand({ TableName: this.table, Key: this.key(op) }));
  }
}

/**
 * Creates the table with the single pk string partition key, on-demand billing and TTL on the ttl attribute.
 * Idempotent. Meant for tests and local development; a production table comes from infrastructure code.
 */
export async function ensureTable(
  client: DynamoDBClient,
  tableName = 'anyonce_records',
): Promise<void> {
  try {
    await client.send(
      new CreateTableCommand({
        TableName: tableName,
        BillingMode: 'PAY_PER_REQUEST',
        AttributeDefinitions: [{ AttributeName: 'pk', AttributeType: 'S' }],
        KeySchema: [{ AttributeName: 'pk', KeyType: 'HASH' }],
      }),
    );
  } catch (error) {
    if (!(error instanceof ResourceInUseException)) throw error;
  }
  let active = false;
  for (let i = 0; i < 50; i++) {
    const described = await client.send(new DescribeTableCommand({ TableName: tableName }));
    if (described.Table?.TableStatus === 'ACTIVE') {
      active = true;
      break;
    }
    await new Promise((r) => setTimeout(r, 200));
  }
  if (!active)
    throw new Error(
      `anyonce: dynamodb table ${tableName} did not become ACTIVE within 10 seconds; create it out of band or retry once the table finishes provisioning`,
    );
  try {
    await client.send(
      new UpdateTimeToLiveCommand({
        TableName: tableName,
        TimeToLiveSpecification: { AttributeName: 'ttl', Enabled: true },
      }),
    );
  } catch (error) {
    // Already enabled: DynamoDB answers with a ValidationException naming the current state.
    if (!(error instanceof Error && /TimeToLive is already enabled/i.test(error.message)))
      throw error;
  }
}

File: go/store/dynamodb/dynamodb.go
// Package dynamodb is the DynamoDB store (REQ-ST-DDB-1): begin, complete and abandon are each one conditional
// write, and a refused write carries the old item back through ReturnValuesOnConditionCheckFailure, so no second
// read is needed to classify it.
package dynamodb

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsdynamodb "github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
	"github.com/sns45/anyonce/go/anyonce"
	"github.com/sns45/anyonce/go/store/internal/rowcodec"
)

// MaxResultBytes is Q20: a DynamoDB item is capped at 400 KB, so bodies above this are stored in the omitted form.
const MaxResultBytes = 307_200

// DefaultTable is the table name New uses when Options.Table is empty.
const DefaultTable = "anyonce_records"

// DefaultNativeTTLGrace is added to the native ttl attribute when Options.NativeTTLGrace is zero.
const DefaultNativeTTLGrace = 60 * time.Second

const (
	inFlight  = string(anyonce.StateInFlight)
	completed = string(anyonce.StateCompleted)
)

// KeySeparator is the unit separator: the scope and the key are joined with it into the single partition key
// pk, so one composed key stays unambiguous (Q22).
const KeySeparator = string(rune(31))

// ItemKey is the item key an operation maps to: the scope and the key joined by KeySeparator (Q22).
func ItemKey(scope, k string) string { return scope + KeySeparator + k }

// Options configure the store. Table defaults to DefaultTable, whose single partition key pk holds the scope
// and the key (Q22); there is no sort key.
// NativeTTLGrace is added to the native ttl attribute so a late complete from the previous fence holder still
// finds its row; it defaults to DefaultNativeTTLGrace.
type Options struct {
	Table          string
	NativeTTLGrace time.Duration
}

// Store is the DynamoDB implementation of anyonce.Store.
type Store struct {
	client *awsdynamodb.Client
	table  string
	grace  time.Duration
}

// New builds a store over an existing DynamoDB client.
func New(client *awsdynamodb.Client, opts Options) *Store {
	s := &Store{client: client, table: opts.Table, grace: opts.NativeTTLGrace}
	if s.table == "" {
		s.table = DefaultTable
	}
	if s.grace == 0 {
		s.grace = DefaultNativeTTLGrace
	}
	return s
}

func key(scope, k string) map[string]types.AttributeValue {
	return map[string]types.AttributeValue{
		"pk": &types.AttributeValueMemberS{Value: ItemKey(scope, k)},
	}
}

// splitKey undoes ItemKey, splitting on the first separator: a scope never contains one, a key may.
func splitKey(pk string) (scope, k string) {
	at := strings.Index(pk, KeySeparator)
	if at < 0 {
		return pk, ""
	}
	return pk[:at], pk[at+len(KeySeparator):]
}

func num(value int64) types.AttributeValue {
	return &types.AttributeValueMemberN{Value: strconv.FormatInt(value, 10)}
}

func str(value string) types.AttributeValue {
	return &types.AttributeValueMemberS{Value: value}
}

func readNum(item map[string]types.AttributeValue, name string) int64 {
	if attr, ok := item[name].(*types.AttributeValueMemberN); ok {
		if parsed, err := strconv.ParseInt(attr.Value, 10, 64); err == nil {
			return parsed
		}
	}
	return 0
}

func readStr(item map[string]types.AttributeValue, name string) string {
	if attr, ok := item[name].(*types.AttributeValueMemberS); ok {
		return attr.Value
	}
	return ""
}

func itemToRow(item map[string]types.AttributeValue) rowcodec.Row {
	scope, k := splitKey(readStr(item, "pk"))
	row := rowcodec.Row{
		Scope:         scope,
		Key:           k,
		Fingerprint:   readStr(item, "fingerprint"),
		State:         readStr(item, "state"),
		Fence:         readNum(item, "fence"),
		LeaseUntil:    readNum(item, "lease_until"),
		CreatedAt:     readNum(item, "created_at"),
		ExpiresAt:     readNum(item, "expires_at"),
		ResultOmitted: readNum(item, "result_omitted"),
	}
	if row.State == "" {
		row.State = inFlight
	}
	if attr, ok := item["result_meta"].(*types.AttributeValueMemberS); ok {
		meta := attr.Value
		row.ResultMeta = &meta
	}
	if attr, ok := item["result_body"].(*types.AttributeValueMemberB); ok {
		row.ResultBody = attr.Value
	}
	return row
}

// refusal returns the old item a conditional write carried back, and whether err was such a refusal.
func refusal(err error) (map[string]types.AttributeValue, bool) {
	var failed *types.ConditionalCheckFailedException
	if !errors.As(err, &failed) {
		return nil, false
	}
	return failed.Item, true
}

// unclassifiableRefusal is the error for a refusal that carried no old item back, which leaves Begin with
// nothing to classify. It names the option that has to be honoured, as the TypeScript store's message does.
func unclassifiableRefusal(err error) error {
	return fmt.Errorf("dynamodb: begin was refused without returning the old item, so the refusal cannot be "+
		"classified; the endpoint must honour ReturnValuesOnConditionCheckFailure (DynamoDB Local 2.x or later, "+
		"or a live table): %w", err)
}

// Begin claims the operation with one conditional UpdateItem. A refusal is classified from the old item the
// write returned: TTL expiry first, then fingerprint, then state and lease.
func (s *Store) Begin(ctx context.Context, op anyonce.Operation, opts anyonce.BeginOptions) (anyonce.BeginOutcome, error) {
	nowMs := opts.Now.UnixMilli()
	for attempt := 0; attempt < 3; attempt++ {
		// The native attribute is a wall clock deadline in seconds; every logical field comes from opts.Now.
		ttlSeconds := time.Now().Add(opts.TTL + s.grace).Unix()
		out, err := s.client.UpdateItem(ctx, &awsdynamodb.UpdateItemInput{
			TableName:           aws.String(s.table),
			Key:                 key(op.Scope, op.Key),
			ConditionExpression: aws.String("attribute_not_exists(pk) OR expires_at <= :now OR (fingerprint = :fp AND #state = :in_flight AND lease_until <= :now)"),
			UpdateExpression: aws.String("SET fingerprint = :fp, #state = :in_flight, fence = if_not_exists(fence, :zero) + :one, " +
				"lease_until = :lease, created_at = :now, expires_at = :exp, #ttl = :ttl, result_omitted = :zero REMOVE result_meta, result_body"),
			ExpressionAttributeNames: map[string]string{"#state": "state", "#ttl": "ttl"},
			ExpressionAttributeValues: map[string]types.AttributeValue{
				":fp":        str(op.Fingerprint),
				":in_flight": str(inFlight),
				":now":       num(nowMs),
				":zero":      num(0),
				":one":       num(1),
				":lease":     num(opts.Now.Add(opts.Lease).UnixMilli()),
				":exp":       num(opts.Now.Add(opts.TTL).UnixMilli()),
				":ttl":       num(ttlSeconds),
			},
			ReturnValues:                        types.ReturnValueAllNew,
			ReturnValuesOnConditionCheckFailure: types.ReturnValuesOnConditionCheckFailureAllOld,
		})
		if err == nil {
			return anyonce.BeginOutcome{Kind: anyonce.BeginAcquired, Fence: readNum(out.Attributes, "fence")}, nil
		}
		item, refused := refusal(err)
		if !refused {
			return anyonce.BeginOutcome{}, fmt.Errorf("dynamodb: begin: %w", err)
		}
		if item == nil {
			return anyonce.BeginOutcome{}, unclassifiableRefusal(err)
		}
		row := itemToRow(item)
		if row.ExpiresAt <= nowMs {
			continue
		}
		record := rowcodec.ToRecord(row)
		if row.Fingerprint != op.Fingerprint {
			return anyonce.BeginOutcome{Kind: anyonce.BeginMismatch, Record: &record}, nil
		}
		if row.State == completed {
			return anyonce.BeginOutcome{Kind: anyonce.BeginCompleted, Record: &record}, nil
		}
		if row.LeaseUntil > nowMs {
			return anyonce.BeginOutcome{Kind: anyonce.BeginInFlight, LeaseUntil: time.UnixMilli(row.LeaseUntil).UTC()}, nil
		}
	}
	return anyonce.BeginOutcome{}, errors.New("dynamodb: begin could not settle after three attempts")
}

// Complete stores the result if the fence still holds and the row is live and in flight.
func (s *Store) Complete(ctx context.Context, op anyonce.Operation, fence int64, result anyonce.StoredResult, now time.Time) (anyonce.CompleteStatus, error) {
	if !result.Omitted && len(result.Body) > MaxResultBytes {
		return "", fmt.Errorf("dynamodb: cannot store a %d byte body because an item is capped at 400 KB (Q20); "+
			"set Policy.MaxResultBytes to at most %d so a larger result is stored in the omitted form instead",
			len(result.Body), MaxResultBytes)
	}
	meta, err := rowcodec.EncodeMeta(result)
	if err != nil {
		return "", fmt.Errorf("dynamodb: complete: %w", err)
	}
	nowMs := now.UnixMilli()
	omitted := int64(0)
	if result.Omitted {
		omitted = 1
	}
	values := map[string]types.AttributeValue{
		":fence":     num(fence),
		":now":       num(nowMs),
		":in_flight": str(inFlight),
		":completed": str(completed),
		":meta":      str(meta),
		":om":        num(omitted),
	}
	update := "SET #state = :completed, result_meta = :meta, result_omitted = :om"
	if !result.Omitted && result.Body != nil {
		values[":body"] = &types.AttributeValueMemberB{Value: result.Body}
		update += ", result_body = :body"
	} else {
		update += " REMOVE result_body"
	}
	_, err = s.client.UpdateItem(ctx, &awsdynamodb.UpdateItemInput{
		TableName:                           aws.String(s.table),
		Key:                                 key(op.Scope, op.Key),
		ConditionExpression:                 aws.String("fence = :fence AND expires_at > :now AND #state = :in_flight"),
		UpdateExpression:                    aws.String(update),
		ExpressionAttributeNames:            map[string]string{"#state": "state"},
		ExpressionAttributeValues:           values,
		ReturnValuesOnConditionCheckFailure: types.ReturnValuesOnConditionCheckFailureAllOld,
	})
	if err == nil {
		return anyonce.CompleteOK, nil
	}
	item, refused := refusal(err)
	if !refused {
		return "", fmt.Errorf("dynamodb: complete: %w", err)
	}
	if item == nil {
		return anyonce.CompleteNotFound, nil
	}
	row := itemToRow(item)
	switch {
	case row.ExpiresAt <= nowMs:
		return anyonce.CompleteNotFound, nil
	case row.Fence != fence:
		return anyonce.CompleteStaleFence, nil
	default:
		// The row is already completed at this fence: completing twice is idempotent.
		return anyonce.CompleteOK, nil
	}
}

// Abandon deletes the in-flight row if the fence still holds.
func (s *Store) Abandon(ctx context.Context, op anyonce.Operation, fence int64) (anyonce.CompleteStatus, error) {
	_, err := s.client.DeleteItem(ctx, &awsdynamodb.DeleteItemInput{
		TableName:           aws.String(s.table),
		Key:                 key(op.Scope, op.Key),
		ConditionExpression: aws.String("fence = :fence AND #state = :in_flight"),
		ExpressionAttributeNames: map[string]string{
			"#state": "state",
		},
		ExpressionAttributeValues: map[string]types.AttributeValue{
			":fence":     num(fence),
			":in_flight": str(inFlight),
		},
		ReturnValuesOnConditionCheckFailure: types.ReturnValuesOnConditionCheckFailureAllOld,
	})
	if err == nil {
		return anyonce.CompleteOK, nil
	}
	item, refused := refusal(err)
	if !refused {
		return "", fmt.Errorf("dynamodb: abandon: %w", err)
	}
	if item == nil {
		return anyonce.CompleteNotFound, nil
	}
	row := itemToRow(item)
	if row.State != inFlight {
		return anyonce.CompleteNotFound, nil
	}
	if row.Fence == fence {
		return anyonce.CompleteNotFound, nil
	}
	return anyonce.CompleteStaleFence, nil
}

// Get reads the record with a strongly consistent read; a row past its expires_at counts as absent.
func (s *Store) Get(ctx context.Context, scope, k string, now time.Time) (*anyonce.Record, error) {
	out, err := s.client.GetItem(ctx, &awsdynamodb.GetItemInput{
		TableName:      aws.String(s.table),
		Key:            key(scope, k),
		ConsistentRead: aws.Bool(true),
	})
	if err != nil {
		return nil, fmt.Errorf("dynamodb: get: %w", err)
	}
	if out.Item == nil {
		return nil, nil
	}
	row := itemToRow(out.Item)
	if row.ExpiresAt <= now.UnixMilli() {
		return nil, nil
	}
	record := rowcodec.ToRecord(row)
	return &record, nil
}

// Purge is a no-op that returns 0: the native ttl attribute sweeps expired items (REQ-ST-DDB-1).
func (s *Store) Purge(context.Context, time.Time) (int, error) {
	return 0, nil
}

// PhysicallyRemove deletes the item unconditionally, which is what a TTL sweep does. Test-only.
func (s *Store) PhysicallyRemove(ctx context.Context, scope, k string) error {
	if _, err := s.client.DeleteItem(ctx, &awsdynamodb.DeleteItemInput{TableName: aws.String(s.table), Key: key(scope, k)}); err != nil {
		return fmt.Errorf("dynamodb: physically remove: %w", err)
	}
	return nil
}

// EnsureTable creates the table with the single pk string partition key and on-demand billing, waits for it to
// become active, then enables TTL on the ttl attribute. It is idempotent. Meant for tests and local
// development; a production table comes from infrastructure code.
func EnsureTable(ctx context.Context, client *awsdynamodb.Client, table string) error {
	if table == "" {
		table = DefaultTable
	}
	_, err := client.CreateTable(ctx, &awsdynamodb.CreateTableInput{
		TableName:   aws.String(table),
		BillingMode: types.BillingModePayPerRequest,
		AttributeDefinitions: []types.AttributeDefinition{
			{AttributeName: aws.String("pk"), AttributeType: types.ScalarAttributeTypeS},
		},
		KeySchema: []types.KeySchemaElement{
			{AttributeName: aws.String("pk"), KeyType: types.KeyTypeHash},
		},
	})
	if err != nil {
		var inUse *types.ResourceInUseException
		if !errors.As(err, &inUse) {
			return fmt.Errorf("dynamodb: create table %s: %w", table, err)
		}
	}
	waiter := awsdynamodb.NewTableExistsWaiter(client)
	if err := waiter.Wait(ctx, &awsdynamodb.DescribeTableInput{TableName: aws.String(table)}, 30*time.Second); err != nil {
		return fmt.Errorf("dynamodb: wait for table %s: %w", table, err)
	}
	_, err = client.UpdateTimeToLive(ctx, &awsdynamodb.UpdateTimeToLiveInput{
		TableName:               aws.String(table),
		TimeToLiveSpecification: &types.TimeToLiveSpecification{AttributeName: aws.String("ttl"), Enabled: aws.Bool(true)},
	})
	if err != nil {
		// Already enabled: DynamoDB answers with a ValidationException naming the current state. The SDK models
		// no Go type for it, so the message is what identifies it, as in the TypeScript store.
		if strings.Contains(strings.ToLower(err.Error()), "timetolive is already enabled") {
			return nil
		}
		return fmt.Errorf("dynamodb: enable ttl on %s: %w", table, err)
	}
	return nil
}
